### DISCUSSION TITLE
Simple .AddSingleton examples?

### DISCUSSION THREAD

**Discussion by dcowman99** [Gremlinq/ExRam.Gremlinq#536] [Q&A]
Hello I'm trying to find examples of the services.AddSingleton pattern for typical .NET dependency injection.  I want to see the various configuration options but can't find any relevant documentation on different ways to connect Gremlinq to CosmosDB.   Where are you guys finding those examples?

**Comment by danielcweber** [comment]
They're here: https://github.com/ExRam/ExRam.Gremlinq.Samples. The link is also on the readme page of this repo.

**Reply by dcowman99** [reply]
Thanks for your response, but the AspNet sample only shows one way to create the connection:

`                .AddGremlinq(setup => setup
                    .UseCosmosDb<Vertex, Edge>(x => x.PartitionKey))`

Is this the only option?  Isn't there a way to itemize each of the settings here in a more explicit way?  Something like this:

`services.AddGremlinq(setup => setup.UseCosmosDb<Vertex, Edge>(x => x.PartitionKey))
                    .At(new Uri("wss://your_gremlin_endpoint.gremlin.cosmos.azure.com:443/"))
                    .OnDatabase("your database name")
                    .OnGraph("your graph name")
                    .AuthenticateBy("your auth key")`

The above code doesn't work obviously but that's the type of example I'm looking for.  The azure portal does not have a way to define nested configuration settings such as the "CosmosDb" section in the example appsettings.json.  We need to define them each individually.

Thanks for your help.

**Comment by danielcweber** [comment]
You mean setting up an IGremlinQuerySource like [this](https://github.com/ExRam/ExRam.Gremlinq.Samples/blob/main/ExRam.Gremlinq.Samples.Console/Program.cs#L23) and then registering the IGremlinQuerySource by AddSingleton? Should be all you need.

**Reply by dcowman99** [reply]
Getting closer.  Tried using the "var gremlinQuerySource = g" example with:
services.AddSingleton<IGremlinQuerySource, GremlinQuerySource>();

AddSingleton doesn't work because the GremlinQuerySource type is static.

Next I tried the following code.  The .UseCosmosDb section has a complie error "Delegate Func<ICosmosDbConfigurator, IConfiguration, ICosmosDbConfigurator> does not take 1 arguments.

```
services.AddGremlinq(a => new GremlinqSetup(services)
                .ConfigureEnvironment(env => env //We call ConfigureEnvironment twice so that the logger is set on the environment from now on.
                    .UseLogger(LoggerFactory
                        .Create(builder => builder
                            .AddFilter(__ => true)
                            .AddConsole())
                        .CreateLogger("Queries")))
                .ConfigureEnvironment(env => env
                    .UseModel(GraphModel
                        .FromBaseTypes<Vertex, Edge>(lookup => lookup
                            .IncludeAssembliesOfBaseTypes())
                        //For CosmosDB, we exclude the 'PartitionKey' property from being included in updates.
                        .ConfigureProperties(model => model
                            .ConfigureElement<Vertex>(conf => conf
                                .IgnoreOnUpdate(x => x.pk))))
                    //Disable query logging for a noise free console output.
                    //Enable logging by setting the verbosity to anything but None.
                    .ConfigureOptions(options => options
                        .SetValue(WebSocketGremlinqOptions.QueryLogLogLevel, LogLevel.None)))

                 .UseCosmosDb(configurator => configurator
                    .At(new Uri("wss://your_gremlin_endpoint.gremlin.cosmos.azure.com:443/"))
                    .OnDatabase("your database name")
                    .OnGraph("your graph name")
                    .AuthenticateBy("your auth key")
                    .ConfigureWebSocket(_ => _
                        .ConfigureGremlinClient(client => client
                            .ObserveResultStatusAttributes((requestMessage, statusAttributes) =>
                            {
                                //Uncomment to log request charges for CosmosDB.
                                //if (statusAttributes.TryGetValue("x-ms-total-request-charge", out var requestCharge))
                                //    env.Logger.LogInformation($"Query {requestMessage.RequestId} had a RU charge of {requestCharge}.");
                            }))));
```

**Comment by danielcweber** [comment]
You might have to post a minimum repo of what you're trying to achieve. Also, please not that the samples run on Gremlinq 9.1.0. I am not sure what your setup is.
